cupy/logic/truth.py

```diff
@@ -82,23 +84,41 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
         cupy.ndarray, bool: The values ``ar1[in1d]`` are in ``ar2``.
 
     """
-    # Ravel both arrays, behavior for the first array could be different
     ar1 = ar1.ravel()
     ar2 = ar2.ravel()
-    if ar1.size == 0 or ar2.size == 0:
+
+    # Seems this conditions works nicely
+    if len(ar2) < 10 * len(ar1) ** 0.145:
         if invert:
-            return cupy.ones(ar1.shape, dtype=cupy.bool_)
+            mask = cupy.ones(len(ar1), dtype=cupy.bool_)
+            for a in ar2:
+                mask &= (ar1 != a)
         else:
-            return cupy.zeros(ar1.shape, dtype=cupy.bool_)
-
-    shape = (ar1.size, ar2.size)
-    ar1_broadcast = cupy.broadcast_to(ar1[..., cupy.newaxis], shape)
-    ar2_broadcast = cupy.broadcast_to(ar2, shape)
-    count = (ar1_broadcast == ar2_broadcast).sum(axis=1)
+            mask = cupy.zeros(len(ar1), dtype=cupy.bool_)
+            for a in ar2:
+                mask |= (ar1 == a)
+        return mask
+
+    # Otherwise use sorting
+    if not assume_unique:
+        ar1, rev_idx = cupy.unique(ar1, return_inverse=True)
+        ar2 = cupy.unique(ar2)
+
+    ar = cupy.concatenate((ar1, ar2))
+    order = ar.argsort()  # ok, cupy gives stable sort
+    sar = ar[order]
     if invert:
-        return count == 0
+        bool_ar = (sar[1:] != sar[:-1])
+    else:
+        bool_ar = (sar[1:] == sar[:-1])
+    flag = cupy.concatenate((bool_ar, cupy.asarray([invert])))
+    ret = cupy.empty(ar.shape, dtype=cupy.bool_)
+    ret[order] = flag
+
+    if assume_unique:
+        return ret[:len(ar1)]
     else:
-        return count > 0
+        return ret[rev_idx]
```

Comment:
How about using `cupy.searchsorted`? This implementation is faster than this PR and does not cause any CPU performance degradation.
```
    ar2 = cupy.sort(ar2)
    v1 = cupy.searchsorted(ar2, ar1, 'left')
    v2 = cupy.searchsorted(ar2, ar1, 'right')
    return v1 == v2 if invert else v1 != v2
```

```
in1d                :    CPU:  158.895 us   +/-19.169 (min:  155.257 / max: 1831.730) us     GPU-0:  362.021 us   +/-18.003 (min:  357.376 / max: 1875.968) us
```

Reply:
This is a brilliant piece. To be applied. With this I would ignore the workarounds for `if len(ar2) < 10 * len(ar1) ** 0.145.

Reply:
I pushed new `in1d` with this code. My humble investigations on my laptop P1000 show that this is much faster then cupy if only `N = ar1.size = ar2.size > ~300`. If `N < 300` the worst case is that this is two times slower. Didn't tested other devices and didn't tested the case when `ar1.size != ar2.size`.

Reply:
> How about using `cupy.searchsorted`? This implementation is faster than this PR and does not cause any CPU performance degradation.
> 
> ```
>     ar2 = cupy.sort(ar2)
>     v1 = cupy.searchsorted(ar2, ar1, 'left')
>     v2 = cupy.searchsorted(ar2, ar1, 'right')
>     return v1 == v2 if invert else v1 != v2
> ```
> 
> ```
> in1d                :    CPU:  158.895 us   +/-19.169 (min:  155.257 / max: 1831.730) us     GPU-0:  362.021 us   +/-18.003 (min:  357.376 / max: 1875.968) us
> ```

.isin() on my laptop on CPU is **6 times slower** than stock numpy on (200000,100) array..
```

def in1d(ar1, ar2, assume_unique=False, invert=False):

	ar1 = ar1.ravel()
	ar2 = ar2.ravel()
	if ar1.size == 0 or ar2.size == 0:
		if invert:
			return np.ones(ar1.shape, dtype=np.bool)
		else:
			return np.zeros(ar1.shape, dtype=np.bool)
	ar2 = np.sort(ar2)
	v1 = np.searchsorted(ar2, ar1, 'left')
	v2 = np.searchsorted(ar2, ar1, 'right')
	return v1 == v2 if invert else v1 != v2

def isin(element, test_elements, assume_unique=False, invert=False):
	return in1d(element, test_elements, assume_unique=assume_unique, invert=invert).reshape(element.shape)
```